From eec7d4055bd9eac93a0107e507c6e1c18d5033a3 Mon Sep 17 00:00:00 2001
From: Victor Bo <bvoid@yandex.ru>
Date: Mon, 23 Dec 2019 14:42:14 +0200
Subject: [PATCH 04/73] Move IMEI information to Sim status

  * from
  https://github.com/Havoc-OS/android_packages_apps_Settings/commit/3741d37c0639883cdc6855bb371fce259ffc0775#diff-8e0314d48f88bb73a3b9ebce6165548d

	modified:   res/layout/dialog_sim_status.xml
	modified:   res/values/strings.xml
	modified:   res/xml/my_device_info.xml
	modified:   src/com/android/settings/deviceinfo/aboutphone/MyDeviceInfoFragment.java
	modified:   src/com/android/settings/deviceinfo/simstatus/SimStatusDialogController.java
---
 res/layout/dialog_sim_status.xml              | 80 +++++++++++++++++
 res/values/strings.xml                        |  2 +-
 res/xml/my_device_info.xml                    | 59 ------------
 .../aboutphone/MyDeviceInfoFragment.java      | 10 ---
 .../simstatus/SimStatusDialogController.java  | 89 +++++++++++++++++++
 5 files changed, 170 insertions(+), 70 deletions(-)

diff --git a/res/layout/dialog_sim_status.xml b/res/layout/dialog_sim_status.xml
index 27d12a8cb1..9383499689 100644
--- a/res/layout/dialog_sim_status.xml
+++ b/res/layout/dialog_sim_status.xml
@@ -180,5 +180,85 @@
             android:layout_width="match_parent"
             android:layout_height="wrap_content"
             android:text="@string/device_info_not_available"/>
+
+        <!-- These items are only for CDMA phones -->
+        <LinearLayout
+            android:id="@+id/cdma_settings"
+            android:orientation="vertical"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content">
+            <TextView
+                style="@style/device_info_dialog_label"
+                android:id="@+id/min_number_label"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:text="@string/status_min_number"/>
+            <TextView
+                style="@style/device_info_dialog_value"
+                android:id="@+id/min_number_value"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:text="@string/device_info_not_available"/>
+
+            <TextView
+                style="@style/device_info_dialog_label"
+                android:id="@+id/prl_version_label"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:text="@string/status_prl_version"/>
+            <TextView
+                style="@style/device_info_dialog_value"
+                android:id="@+id/prl_version_value"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:text="@string/device_info_not_available"/>
+
+            <TextView
+                style="@style/device_info_dialog_label"
+                android:id="@+id/meid_number_label"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:text="@string/status_meid_number"/>
+            <TextView
+                style="@style/device_info_dialog_value"
+                android:id="@+id/meid_number_value"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:text="@string/device_info_not_available"/>
+
+        </LinearLayout>
+
+        <!-- These items are for GSM and LTE enabled CDMA phones -->
+        <LinearLayout
+            android:id="@+id/gsm_settings"
+            android:orientation="vertical"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content">
+            <TextView
+                style="@style/device_info_dialog_label"
+                android:id="@+id/imei_label"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:text="@string/status_imei"/>
+            <TextView
+                style="@style/device_info_dialog_value"
+                android:id="@+id/imei_value"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:text="@string/device_info_not_available"/>
+
+            <TextView
+                style="@style/device_info_dialog_label"
+                android:id="@+id/imei_sv_label"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:text="@string/status_imei_sv"/>
+            <TextView
+                style="@style/device_info_dialog_value"
+                android:id="@+id/imei_sv_value"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:text="@string/device_info_not_available"/>
+        </LinearLayout>
     </LinearLayout>
 </ScrollView>
diff --git a/res/values/strings.xml b/res/values/strings.xml
index 811a44c7c5..ed9aa5b6ba 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -7486,7 +7486,7 @@
     <!-- SIM status title  [CHAR LIMIT=40] -->
     <string name="sim_status_title">SIM status</string>
     <!-- SIM status title  [CHAR LIMIT=40] -->
-    <string name="sim_status_title_sim_slot">SIM status (sim slot %1$d)</string>
+    <string name="sim_status_title_sim_slot">SIM status (Slot %1$d)</string>
     <!-- Title for call back. [CHAR LIMIT=60] -->
     <string name="sim_call_back_title">Call back from default SIM</string>
     <!-- Title for outgoing back. [CHAR LIMIT=60] -->
diff --git a/res/xml/my_device_info.xml b/res/xml/my_device_info.xml
index 5db2123c65..85b1a8f89a 100644
--- a/res/xml/my_device_info.xml
+++ b/res/xml/my_device_info.xml
@@ -45,16 +45,6 @@
         android:summary="@string/summary_placeholder"
         settings:controller="com.android.settings.deviceinfo.BrandedAccountPreferenceController"/>
 
-    <!-- Phone number -->
-    <Preference
-        android:key="phone_number"
-        android:order="3"
-        android:title="@string/status_number"
-        android:summary="@string/summary_placeholder"
-        android:selectable="false"
-        settings:controller="com.android.settings.deviceinfo.PhoneNumberPreferenceController"
-        settings:enableCopying="true"/>
-
     <Preference
         android:key="emergency_info"
         android:order="14"
@@ -62,30 +52,6 @@
         android:summary="@string/summary_placeholder"
         settings:controller="com.android.settings.accounts.EmergencyInfoPreferenceController"/>
 
-    <!-- Legal information -->
-    <Preference
-        android:key="legal_container"
-        android:order="15"
-        android:title="@string/legal_information"
-        android:fragment="com.android.settings.LegalSettings"
-        settings:allowDividerAbove="true"/>
-
-    <!-- Regulatory labels -->
-    <Preference
-        android:key="regulatory_info"
-        android:order="16"
-        android:title="@string/regulatory_labels">
-        <intent android:action="android.settings.SHOW_REGULATORY_INFO"/>
-    </Preference>
-
-    <!-- Safety & regulatory manual -->
-    <Preference
-        android:key="safety_info"
-        android:order="17"
-        android:title="@string/safety_and_regulatory_info">
-        <intent android:action="android.settings.SHOW_SAFETY_AND_REGULATORY_INFO"/>
-    </Preference>
-
     <!-- SIM status -->
     <Preference
         android:key="sim_status"
@@ -107,16 +73,6 @@
         settings:keywords="@string/keywords_model_and_hardware"
         settings:controller="com.android.settings.deviceinfo.HardwareInfoPreferenceController"/>
 
-    <!-- IMEI -->
-    <Preference
-        android:key="imei_info"
-        android:order="32"
-        android:title="@string/status_imei"
-        android:summary="@string/summary_placeholder"
-        settings:keywords="@string/keywords_imei_info"
-        settings:enableCopying="true"
-        settings:controller="com.android.settings.deviceinfo.imei.ImeiInfoPreferenceController"/>
-
     <!-- Android version -->
     <Preference
         android:key="firmware_version"
@@ -171,21 +127,6 @@
         android:summary="@string/summary_placeholder"
         android:selectable="false"/>
 
-    <!-- Manual -->
-    <Preference
-        android:key="manual"
-        android:order="50"
-        android:title="@string/manual">
-        <intent android:action="android.settings.SHOW_MANUAL"/>
-    </Preference>
-
-    <!-- Feedback on the device -->
-    <Preference
-        android:key="device_feedback"
-        android:order="51"
-        android:title="@string/device_feedback"
-        settings:keywords="@string/keywords_device_feedback"/>
-
     <!-- Device FCC equipment id -->
     <Preference
         android:key="fcc_equipment_id"
diff --git a/src/com/android/settings/deviceinfo/aboutphone/MyDeviceInfoFragment.java b/src/com/android/settings/deviceinfo/aboutphone/MyDeviceInfoFragment.java
index 4af5d79c37..f963e9e2ce 100644
--- a/src/com/android/settings/deviceinfo/aboutphone/MyDeviceInfoFragment.java
+++ b/src/com/android/settings/deviceinfo/aboutphone/MyDeviceInfoFragment.java
@@ -32,14 +32,9 @@ import com.android.settings.deviceinfo.BluetoothAddressPreferenceController;
 import com.android.settings.deviceinfo.BuildNumberPreferenceController;
 import com.android.settings.deviceinfo.DeviceNamePreferenceController;
 import com.android.settings.deviceinfo.FccEquipmentIdPreferenceController;
-import com.android.settings.deviceinfo.FeedbackPreferenceController;
 import com.android.settings.deviceinfo.IpAddressPreferenceController;
-import com.android.settings.deviceinfo.ManualPreferenceController;
-import com.android.settings.deviceinfo.RegulatoryInfoPreferenceController;
-import com.android.settings.deviceinfo.SafetyInfoPreferenceController;
 import com.android.settings.deviceinfo.UptimePreferenceController;
 import com.android.settings.deviceinfo.WifiMacAddressPreferenceController;
-import com.android.settings.deviceinfo.imei.ImeiInfoPreferenceController;
 import com.android.settings.deviceinfo.simstatus.SimStatusPreferenceController;
 import com.android.settings.search.BaseSearchIndexProvider;
 import com.android.settings.widget.EntityHeaderController;
@@ -73,7 +68,6 @@ public class MyDeviceInfoFragment extends DashboardFragment
     @Override
     public void onAttach(Context context) {
         super.onAttach(context);
-        use(ImeiInfoPreferenceController.class).setHost(this /* parent */);
         use(DeviceNamePreferenceController.class).setHost(this /* parent */);
         mBuildNumberPreferenceController = use(BuildNumberPreferenceController.class);
         mBuildNumberPreferenceController.setHost(this /* parent */);
@@ -107,10 +101,6 @@ public class MyDeviceInfoFragment extends DashboardFragment
         controllers.add(new IpAddressPreferenceController(context, lifecycle));
         controllers.add(new WifiMacAddressPreferenceController(context, lifecycle));
         controllers.add(new BluetoothAddressPreferenceController(context, lifecycle));
-        controllers.add(new RegulatoryInfoPreferenceController(context));
-        controllers.add(new SafetyInfoPreferenceController(context));
-        controllers.add(new ManualPreferenceController(context));
-        controllers.add(new FeedbackPreferenceController(fragment, context));
         controllers.add(new FccEquipmentIdPreferenceController(context));
         controllers.add(new UptimePreferenceController(context, lifecycle));
         return controllers;
diff --git a/src/com/android/settings/deviceinfo/simstatus/SimStatusDialogController.java b/src/com/android/settings/deviceinfo/simstatus/SimStatusDialogController.java
index 9b54e3be32..347528aabb 100644
--- a/src/com/android/settings/deviceinfo/simstatus/SimStatusDialogController.java
+++ b/src/com/android/settings/deviceinfo/simstatus/SimStatusDialogController.java
@@ -49,12 +49,18 @@ import android.telephony.euicc.EuiccManager;
 import android.telephony.ims.ImsException;
 import android.telephony.ims.ImsMmTelManager;
 import android.telephony.ims.ImsReasonInfo;
+import android.text.Spannable;
+import android.text.SpannableStringBuilder;
+import android.text.Spanned;
 import android.text.TextUtils;
+import android.text.style.TtsSpan;
 import android.util.Log;
 
 import androidx.annotation.NonNull;
 import androidx.annotation.VisibleForTesting;
 
+import com.android.internal.telephony.PhoneConstants;
+
 import com.android.settings.R;
 import com.android.settingslib.DeviceInfoUtils;
 import com.android.settingslib.Utils;
@@ -106,6 +112,21 @@ public class SimStatusDialogController implements LifecycleObserver, OnResume, O
     final static int IMS_REGISTRATION_STATE_LABEL_ID = R.id.ims_reg_state_label;
     @VisibleForTesting
     final static int IMS_REGISTRATION_STATE_VALUE_ID = R.id.ims_reg_state_value;
+    @VisibleForTesting
+    static final int ID_PRL_VERSION_VALUE = R.id.prl_version_value;
+    private static final int ID_MIN_NUMBER_LABEL = R.id.min_number_label;
+    @VisibleForTesting
+    static final int ID_MIN_NUMBER_VALUE = R.id.min_number_value;
+    @VisibleForTesting
+    static final int ID_MEID_NUMBER_VALUE = R.id.meid_number_value;
+    @VisibleForTesting
+    static final int ID_IMEI_VALUE = R.id.imei_value;
+    @VisibleForTesting
+    static final int ID_IMEI_SV_VALUE = R.id.imei_sv_value;
+    @VisibleForTesting
+    static final int ID_CDMA_SETTINGS = R.id.cdma_settings;
+    @VisibleForTesting
+    static final int ID_GSM_SETTINGS = R.id.gsm_settings;
 
     @VisibleForTesting
     static final int MAX_PHONE_COUNT_SINGLE_SIM = 1;
@@ -149,6 +170,7 @@ public class SimStatusDialogController implements LifecycleObserver, OnResume, O
     private final SubscriptionManager mSubscriptionManager;
     private final CarrierConfigManager mCarrierConfigManager;
     private final EuiccManager mEuiccManager;
+    private final int mSlotId;
     private final Resources mRes;
     private final Context mContext;
 
@@ -206,6 +228,7 @@ public class SimStatusDialogController implements LifecycleObserver, OnResume, O
     public SimStatusDialogController(@NonNull SimStatusDialogFragment dialog, Lifecycle lifecycle,
             int slotId) {
         mDialog = dialog;
+        mSlotId = slotId;
         mContext = dialog.getContext();
         mSlotIndex = slotId;
         mSubscriptionInfo = getPhoneSubscriptionInfo(slotId);
@@ -248,6 +271,12 @@ public class SimStatusDialogController implements LifecycleObserver, OnResume, O
         updateRoamingStatus(serviceState);
         updateIccidNumber();
         updateImsRegistrationState();
+
+        if (mTelephonyManager.getPhoneType() == TelephonyManager.PHONE_TYPE_CDMA) {
+            updateDialogForCdmaPhone();
+        } else {
+            updateDialogForGsmPhone();
+        }
     }
 
     /**
@@ -703,6 +732,40 @@ public class SimStatusDialogController implements LifecycleObserver, OnResume, O
         imsMmTelMgr.unregisterImsRegistrationCallback(mImsRegStateCallback);
     }
 
+    private void updateDialogForCdmaPhone() {
+        final Resources res = mDialog.getContext().getResources();
+        mDialog.setText(ID_MEID_NUMBER_VALUE, getMeid());
+        mDialog.setText(ID_MIN_NUMBER_VALUE,
+                mSubscriptionInfo != null ? mTelephonyManager.getCdmaMin(
+                        mSubscriptionInfo.getSubscriptionId()) : "");
+
+        if (res.getBoolean(R.bool.config_msid_enable)) {
+            mDialog.setText(ID_MIN_NUMBER_LABEL,
+                    res.getString(R.string.status_msid_number));
+        }
+
+        mDialog.setText(ID_PRL_VERSION_VALUE, getCdmaPrlVersion());
+
+        if (mSubscriptionInfo != null && isCdmaLteEnabled()) {
+            // Show IMEI for LTE device
+            mDialog.setText(ID_IMEI_VALUE,
+                    getTextAsDigits(mTelephonyManager.getImei(mSlotId)));
+            mDialog.setText(ID_IMEI_SV_VALUE,
+                    getTextAsDigits(mTelephonyManager.getDeviceSoftwareVersion(mSlotId)));
+        } else {
+            // device is not GSM/UMTS, do not display GSM/UMTS features
+            mDialog.removeSettingFromScreen(ID_GSM_SETTINGS);
+        }
+    }
+
+    private void updateDialogForGsmPhone() {
+        mDialog.setText(ID_IMEI_VALUE, getTextAsDigits(mTelephonyManager.getImei(mSlotId)));
+        mDialog.setText(ID_IMEI_SV_VALUE,
+                getTextAsDigits(mTelephonyManager.getDeviceSoftwareVersion(mSlotId)));
+        // device is not CDMA, do not display CDMA features
+        mDialog.removeSettingFromScreen(ID_CDMA_SETTINGS);
+    }
+
     private SubscriptionInfo getPhoneSubscriptionInfo(int slotId) {
         return SubscriptionManager.from(mContext).getActiveSubscriptionInfoForSimSlotIndex(slotId);
     }
@@ -818,4 +881,30 @@ public class SimStatusDialogController implements LifecycleObserver, OnResume, O
                 return "UNKNOWN";
         }
     }
+
+    @VisibleForTesting
+    String getCdmaPrlVersion() {
+        return mTelephonyManager.getCdmaPrlVersion();
+    }
+
+    @VisibleForTesting
+    boolean isCdmaLteEnabled() {
+        return mTelephonyManager.getLteOnCdmaMode(mSubscriptionInfo.getSubscriptionId())
+                == PhoneConstants.LTE_ON_CDMA_TRUE;
+    }
+
+    @VisibleForTesting
+    String getMeid() {
+        return mTelephonyManager.getMeid(mSlotId);
+    }
+
+    private static CharSequence getTextAsDigits(CharSequence text) {
+        if (TextUtils.isDigitsOnly(text)) {
+            final Spannable spannable = new SpannableStringBuilder(text);
+            final TtsSpan span = new TtsSpan.DigitsBuilder(text.toString()).build();
+            spannable.setSpan(span, 0, spannable.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
+            text = spannable;
+        }
+        return text;
+    }
 }
-- 
2.25.1

